### PR TITLE
[mono][wasm] Move some runtime code to a separate runtime.h/runtime.c…

### PR DESCRIPTION
**Pull request by vargaz**
… so it can be shared with wasi.

### CODE REVIEW

**Comment by vargaz** [comment]
The WasmBuildTests failures are relevant.

**Review by pavelsavara — APPROVED** [review]
I like the C part. The build changes would be good to coordinate with https://github.com/dotnet/runtime/pull/95775

**Comment by pavelsavara** [comment]
In `src\mono\wasm\Wasm.Build.Tests\NativeRebuildTests\ReferenceNewAssemblyRebuildTest.cs` line 55 there are expectations on which file was touched during native rebuild. 

[Log](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-95834-merge-2e223b7da2b8435597/Workloads-Wasm.Build.NativeRebuild.Tests.ReferenceNewAssemblyRebuildTest/1/console.86ebfd0c.log?helixlogtype=result)

My guess is that you need to add the new file to 
https://github.com/dotnet/runtime/blob/6de7549ae1872009baade9091be76071f955c47d/src/mono/wasm/Wasm.Build.Tests/ProjectProviderBase.cs#L284-L294

cc @radical

**Comment by vargaz** [comment]
> In `src\mono\wasm\Wasm.Build.Tests\NativeRebuildTests\ReferenceNewAssemblyRebuildTest.cs` line 55 there are expectations on which file was touched during native rebuild.
> 
> [Log](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-95834-merge-2e223b7da2b8435597/Workloads-Wasm.Build.NativeRebuild.Tests.ReferenceNewAssemblyRebuildTest/1/console.86ebfd0c.log?helixlogtype=result)
> 
> My guess is that you need to add the new file to
> 
> https://github.com/dotnet/runtime/blob/6de7549ae1872009baade9091be76071f955c47d/src/mono/wasm/Wasm.Build.Tests/ProjectProviderBase.cs#L284-L294
> 
> cc @radical

Its added, the error complains that driver.o was rebuilt/not rebuilt.

**Comment by radical** [comment]
@vargaz my PR has merged

File: src/mono/wasm/build/WasmApp.targets [hunk 1]
```diff
@@ -336,6 +336,8 @@
                                Dependencies="$(_WasmPInvokeHPath);$(_WasmPInvokeTablePath)" />
       <_WasmRuntimePackSrcFile Include="$(_WasmRuntimePackSrcDir)driver.c"
                                Dependencies="@(_DriverCDependencies)" />
+      <_WasmRuntimePackSrcFile Include="$(_WasmRuntimePackSrcDir)runtime.c"
+                               Dependencies="@(_DriverCDependencies)" />
```
Comment (radical) [inline src/mono/wasm/build/WasmApp.targets:340]: Any other dependencies to add for this?

File: src/mono/wasm/Wasm.Build.Tests/ProjectProviderBase.cs [hunk 2]
```diff
@@ -282,7 +282,7 @@ public static string FindSubDirIgnoringCase(string parentDir, string dirName)
         List<string> files = new()
         {
             Path.Combine(paths.BinDir, "publish", $"{buildArgs.ProjectName}.dll"),
-            Path.Combine(paths.ObjWasmDir, "driver.o"),
```
Comment (radical) [inline src/mono/wasm/Wasm.Build.Tests/ProjectProviderBase.cs:285]: We shouldn't drop this. If this no longer depends on assemblies changing then the tests can expect that now, and expect `runtime.o` to change when assemblies change.

File: src/mono/wasm/build/WasmApp.targets (hunk 1 above)
Comment (vargaz) [reply src/mono/wasm/build/WasmApp.targets:340]: It contains code from driver.c so the same deps should suffice. Probably driver.c needs less deps now.

**Pull request merged**